test/lit_tests/raising/llvm_to_affine_access.mlir

```diff
@@ -0,0 +1,66 @@
+// RUN: enzymexlamlir-opt %s --pass-pipeline="builtin.module(llvm-to-affine-access)" | FileCheck %s
+
+func.func @test_load_store_conversion(%arg0: !llvm.ptr<1>, %idx: i64) {
+  %0 = llvm.getelementptr inbounds %arg0[%idx] : (!llvm.ptr<1>, i64) -> !llvm.ptr<1>, i64
+  %1 = llvm.load %0 {alignment = 1 : i64} : !llvm.ptr<1> -> i64
+  %2 = llvm.mul %1, %1 : i64
+
+  llvm.store %2, %0 {alignment = 1 : i64} : i64, !llvm.ptr<1>
+
+  return
+}
+
+// CHECK-LABEL: func @test_load_store_conversion
+// CHECK-SAME: %[[ARG0:.*]]: !llvm.ptr<1>
+// CHECK-SAME: %[[ARG1:.*]]: i64
+// CHECK: %[[MEMREF:.*]] = "enzymexla.ataddr"(%[[ARG0]]) {{.*}} memref<?xi8, 1>
+// CHECK: %[[IDX:.*]] = arith.index_cast %[[ARG1]]
+// CHECK: affine.vector_load %[[MEMREF]][symbol(%[[IDX]]) * 8] {{.*}} vector<8xi8>
+// CHECK: affine.vector_store
+
+// -----
+
+func.func @test_multidim_load_store(%arg0: !llvm.ptr<1>, %idx1: i64, %idx2: i64) {
+  %c1 = llvm.mlir.constant(1 : index) : i64
+  %ptr = llvm.getelementptr %arg0[%idx1, %idx2] : (!llvm.ptr<1>, i64, i64) -> !llvm.ptr<1>, !llvm.array<8 x i64>
+  %val = llvm.load %ptr : !llvm.ptr<1> -> i64
+
+  %idx1p1 = llvm.add %idx1, %c1 : i64
+  %idx2p1 = llvm.add %idx2, %c1 : i64
+
+  %ptr_str = llvm.getelementptr %arg0[%idx1p1, %idx2p1] : (!llvm.ptr<1>, i64, i64) -> !llvm.ptr<1>, !llvm.array<8 x i64>
+  llvm.store %val, %ptr_str : i64, !llvm.ptr<1>
+
+  return
+}
+
+// CHECK-LABEL: func @test_multidim_load_store
+// CHECK-SAME: %[[ARG0:.*]]: !llvm.ptr<1>,
+// CHECK-SAME: %[[ARG1:.*]]: i64,
+// CHECK-SAME: %[[ARG2:.*]]: i64
+// CHECK: %[[MEMREF:.*]] = "enzymexla.ataddr"(%[[ARG0]]) {{.*}} memref<?xi8, 1>
+// CHECK-DAG: %[[IDX1:.*]] = arith.index_cast %[[ARG1]]
+// CHECK-DAG: %[[IDX2:.*]] = arith.index_cast %[[ARG2]]
+// CHECK: affine.vector_load %[[MEMREF]][symbol(%[[IDX1]]) * 64 + symbol(%[[IDX2]]) * 8] {{.*}} vector<8xi8>
+// CHECK: affine.vector_store
+
+// -----
+
+func.func @test_struct_access(%arg0: !llvm.ptr) {
+  %ptr = llvm.getelementptr %arg0[0, 0] : (!llvm.ptr) -> !llvm.ptr, !llvm.struct<(i64)>
```

Comment:
I feel like it is better to apply this pass after converting all of the arguments to memrefs, that way we know the multidimensional sizes and underlying types

Reply:
Do we ever get multidimensional arrays in our compilation pipeline?

I can see a problem with correctly handling that since all of the loads inside will be with a linear index so e.g. 

```
f(%ptr : !llvm.ptr) {
  ...
  load %j * 4 + %i
  
```
should become this
``` 
f(%ptr : !memref<i8x10x4>) {
  ...
  affine.load[%j][%i]
```
But that is legal only if we can prove that 0 <= %i < 4

Otherwise it needs to be 
```
  affine.load[%j + %i / 4][%i mod 4]
```

I guess we should generate the second one and try to use available information to simplify it to the first one

Reply:
We can also run it before converting the args to memref but we should have a simplification/pattern which goes load(pointer2memref(memref2pointer(memref))) to load(memref) and adjusts the indices appropriately.

Reply:
For TPU I think we'll need to recover the shapes and element types anyway, how hard is it to put that logic into this pass vs. create a separate pass that converts arguments to memref with the correct shapes + raises load/store of arguments to memref.load/store?